cluster detection optimization

JohnnyBraver opened the pull request:
no outlines on the first frame, crashing during render of second
needles to say its unworth of the main branch
*I was not working on this for a while and won't focus like before for a while so I just want my code saved here

JohnnyBraver committed:
wip queue update

no outlines on the first frame, crashing during render of second
needles to say its unworth of the main branch
*I was not working on this for a while and won't focus like before for a while so I just want my code saved here

yoniinoy commented:
missing outlines were due to extreme minimal requirements set for outlining

yoniinoy commented:
the crash seems to happen in the initial for loop of 'FindCluster'that populated the queue with the first cell's neighbors
but not in the first time it's called
nor during the initial render,
the amount of neighbors seems to be unrelated to the crash.

JohnnyBraver committed:
debug progress

nothing major, added a dummy if for debug and changed the outline requirements to something much more reasonable

yoniinoy commented:
the problem seems to do with a mismanaged vector in the initialization of 'FindCluster', pushing into an improperly initialized vector inside the queue

JohnnyBraver committed:
my ouw implementation

pretty big update.
implemented an allocation efficient stack.
globalized widely used arrays and the head of the stack.
some bugs still exist, some clusters that should be outlined are not, probably due to a mistake in the search method,
it runs, but proper speed testing will have to wait until completion and comparable accuracy to the main branch.

JohnnyBraver committed:
minor cleanup

deleted old code that will no longer be used
I hope.
wasn't that complicated anyway so it's not a major loss either way.

JohnnyBraver committed:
bugs fixed

the leak was caused by the global head and input mismanagement, the outline was a similar oversight when reworking the algorithm.
both are fixed and everything seems to function as it should.

JohnnyBraver committed:
stack operation optimization

changed emptiness check to a simple integer counter.
eliminated double additions of the same cell to the stack entirely.
admittedly those changes are only perceivable on the 'lush' setting with any but 'no outlines' enabled but it is a big algorithmic improvement nonetheless

JohnnyBraver committed:
fixed minor display bug

the bitmap wasn't offset to account for the border and the outline was drawn one pixel too soon causing it to miss from one edge and overwrite pixels on another.
both of those issues were fixed, a whole grid cluster now has no visible outline at and any cluster right on the edge has its outline 'missing' as it should.
very light refactor thrown in.

JohnnyBraver committed:
mistake reverted

wrongly affected inactive render method where the bug wasn't observed.

JohnnyBraver committed:
updated cleanup at main exit
